source/vstd/std_specs/hash.rs

```diff
@@ -428,6 +428,234 @@ pub fn ex_hash_map_clear<Key, Value, S>(m: &mut HashMap<Key, Value, S>)
+{
+    HashSet::<Key>::with_capacity(capacity)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_reserve<Key, S>(m: &mut HashSet<Key, S>, additional: usize) where
+    Key: Eq + Hash,
+    S: BuildHasher,
+
+    ensures
+        m@ == old(m)@,
+{
+    m.reserve(additional)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_insert<Key, S>(m: &mut HashSet<Key, S>, k: Key) -> (result: bool) where
+    Key: Eq + Hash,
+    S: BuildHasher,
+
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> {
+            &&& m@ == old(m)@.insert(k)
+            &&& match result {
+                true => old(m)@.contains(k),
+                false => !old(m)@.contains(k),
+            }
+        },
+{
+    m.insert(k)
+}
+
+// The specification for `contains` has a parameter `key: &Q`
+// where you'd expect to find `key: &Key`. This allows for the case
+// that `Key` can be borrowed as something other than `&Key`. For
+// instance, `Box<u32>` can be borrowed as `&u32` and `String` can be
+// borrowed as `&str`, so in those cases `Q` would be `u32` and `str`
+// respectively.
+// To deal with this, we have a specification function that opaquely
+// specifies what it means for a map to contain a borrowed key of type
+// `&Q`. And the postcondition of `contains` just says that its
+// result matches the output of that specification function. But this
+// isn't very helpful by itself, since there's no body to that
+// specification function. So we have special-case axioms that say
+// what this means in two important circumstances: (1) `Key = Q` and
+// (2) `Key = Box<Q>`.
+pub spec fn sets_contains_borrowed_key<Key, Q: ?Sized>(m: Set<Key>, k: &Q) -> bool;
+
+pub broadcast proof fn axiom_sets_contains_deref_key<Q>(m: Set<Q>, k: &Q)
+    ensures
+        #[trigger] sets_contains_borrowed_key::<Q, Q>(m, k) <==> m.contains(*k),
+{
+    admit();
+}
+
+pub broadcast proof fn axiom_sets_contains_box<Q>(m: Set<Box<Q>>, k: &Q)
+    ensures
+        #[trigger] sets_contains_borrowed_key::<Box<Q>, Q>(m, k) <==> m.contains(Box::new(*k)),
+{
+    admit();
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_contains<Key, S, Q>(m: &HashSet<Key, S>, k: &Q) -> (result: bool) where
+    Key: Borrow<Q> + Hash + Eq,
+    Q: Hash + Eq + ?Sized,
+    S: BuildHasher,
+
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> result
+            == sets_contains_borrowed_key(m@, k),
+{
+    m.contains(k)
+}
+
+// The specification for `get` has a parameter `key: &Q` where you'd
+// expect to find `key: &Key`. This allows for the case that `Key` can
+// be borrowed as something other than `&Key`. For instance,
+// `Box<u32>` can be borrowed as `&u32` and `String` can be borrowed
+// as `&str`, so in those cases `Q` would be `u32` and `str`
+// respectively.
+// To deal with this, we have a specification function that opaquely
+// specifies what it means for a map to map a borrowed key of type
```

Comment:
Comment should say "set" instead of "map".

Reply:
Actually, the whole phrase needs to be changed. It specifies what it means for a returned reference to point to an element of a HashSet.